[WIP]: storage volume snapshots

brauner opened the pull request.

Commit:
storage: add new custom volume snapshot API entpoints

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
api: add StorageVolumeSnapshot struct

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
api: add StorageVolumeSnapshotPost struct

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
api: add StorageVolumeSnapshotsPost struct

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
client: add CreateStoragePoolVolumeSnapshot()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

stgraber commented on shared/api/storage_pool_volume_snapshot.go:
```diff
@@ -0,0 +1,27 @@
+
+// StorageVolumeSnapshotsPost represents the fields available for a new LXD storage volume snapshot
+//
+// API extension: storage_api_volume_snapshots
```
What's that Type field?

stgraber replied:
The Type should already be encoded in the URL right, same as the pool name.

brauner replied:
That would make the `type` argument a part of the client function iiuc, right?

stgraber commented on shared/api/storage_pool_volume_snapshot.go:
```diff
@@ -0,0 +1,27 @@
+// StorageVolumeSnapshotsPost represents the fields available for a new LXD storage volume snapshot
+//
+// API extension: storage_api_volume_snapshots
+type StorageVolumeSnapshotsPost struct {
```
Post should only have `Name`

stgraber commented on shared/api/storage_pool_volume_snapshot.go:
```diff
@@ -0,0 +1,27 @@
+// StorageVolumeSnapshotPost represents the fields required to rename/move a LXD storage volume snapshot
+//
+// API extension: storage_api_volume_snapshots
+type StorageVolumeSnapshotPost struct {
```
Post should only have `Name`

stgraber commented on shared/api/storage_pool_volume_snapshot.go:
```diff
@@ -0,0 +1,27 @@
+// StorageVolumeSnapshot represents a LXD storage volume snapshot
+//
+// API extension: storage_api_volume_snapshots
+type StorageVolumeSnapshot struct {
```
This should include quite a few more thing, basically be the same as the storage volume that was snapshotted.
So something like:
 - Name
 - Config
 - Description

Commit:
lxc: add snapshot subcommand to lxc storage

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
api: add "storage_api_volume_snapshots"

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: storagePoolVolumeSnapshotsTypePost()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
db: add StorageVolumeKind to indicate snapshots

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
db: update schema

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
db: add kind argument to StoragePoolVolumeCreate()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: check volume names do not contain /

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: add kind arg to storagePoolVolumeDBCreate

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: use storageValidName() to check snap name

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
api: remove empty import

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: add storagePoolVolumeSnapshotDBCreateInternal()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: use storagePoolVolumeSnapshotDBCreateInternal() in storagePoolVolumeSnapshotsTypePost()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: add StoragePoolVolumeSnapshotCreate()

This adds StoragePoolVolumeSnapshotCreate() to the storage interface. All
functions are not implemented.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
dir: implement StoragePoolVolumeSnapshotCreate()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: use StoragePoolVolumeSnapshotCreate()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
patches: patchStorageApiRenameContainerSnapshotsDir()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: add getStoragePoolVolumeSnapshotMountPoint()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
btrfs: create new "custom-snapshots" subvolume

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
zfs: create new "custom-snapshots" dataset

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
btrfs: add getCustomSnapshotSubvolumePath()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
btrfs: implement StoragePoolVolumeSnapshotCreate()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
[UNTESTED] ceph: implement StoragePoolVolumeSnapshotCreate()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
mock: implement StoragePoolVolumeSnapshotCreate()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
lvm: implement StoragePoolVolumeSnapshotCreate()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
[UNTESTED] zfs: implement StoragePoolVolumeSnapshotCreate()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: add StoragePoolVolumeSnapshotDelete()

This adds StoragePoolVolumeSnapshotDelete() to the storage interface. All
functions are not implemented.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: implement storagePoolVolumeSnapshotTypeDelete()

This implements tha storagePoolVolumeSnapshotTypeDelete() API endpoint.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
dir: implement StoragePoolVolumeSnapshotDelete()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
client: add DeleteStoragePoolVolumeSnapshot()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
lxc: allow to delete custom volume snapshots

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
btrfs: implement StoragePoolVolumeSnapshotDelete()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
[UNTESTED] ceph: implement StoragePoolVolumeSnapshotDelete()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
lvm: implement StoragePoolVolumeSnapshotDelete()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
mock: implement StoragePoolVolumeSnapshotDelete()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
[UNTESTED] zfs: implement StoragePoolVolumeSnapshotDelete()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
client: add GetStoragePoolVolumeSnapshotNames()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
client: add GetStoragePoolVolumeSnapshots()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
db: add StoragePoolVolumeSnapshotsGetType()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: add storagePoolVolumeSnapshotsTypeGet()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: add storagePoolVolumeSnapshotTypeGet()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
client: add GetStoragePoolVolumeSnapshot()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
client: add RenameStoragePoolVolumeSnapshot()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: implement storagePoolVolumeSnapshotTypePost()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: add StoragePoolVolumeSnapshotRename()

This adds StoragePoolVolumeSnapshotRename() to the storage interface. All
functions are not implemented.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
dir: implement StoragePoolVolumeSnapshotRename()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
btrfs: implement StoragePoolVolumeSnapshotRename()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
[UNTESTED] ceph: implement StoragePoolVolumeSnapshotRename()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
[UNTESTED] ceph: fix StoragePoolVolumeRename()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
mock: implement StoragePoolVolumeSnapshotRename()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
lvm: implement StoragePoolVolumeSnapshotRename()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
lvm: fix StoragePoolVolumeRename()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
[UNTESTED] zfs: implement StoragePoolVolumeSnapshotRename()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
lxc: allow storage volume snapshot rename

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
client: add UpdateStoragePoolVolumeSnapshot()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
api: add StorageVolumeSnapshotPut struct

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: add storagePoolVolumeSnapshotTypePut()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: implement storagePoolVolumeSnapshotTypePut()

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: s/snapshots/containers-snapshots/g

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
storage: handle conflicting API handling

Containers need to handle
storage-pools/{pool}/volumes/{type}/{name:.*}
requests to handle snapshots. At the same time custom storage volumes need to
handle
storage-pools/{pool}/volumes/{type}/{name}/snapshots/{snapshotName}. So move
the volume handling into separate commands.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
lxc: enable storage volume snapshot show

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
test: s/snapshots/containers-snapshots/g

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
lxc: enable storage volume snapshot edit

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
lxc: enable storage volume snapshot get

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
i18n: Update translation templates

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

Commit:
api: add Restore field to StorageVolumeSnapshotPut struct

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>
